{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GOAL: alignment skills\n",
    "* BioPython\n",
    "* Python\n",
    "* MUSCLE\n",
    "* algorithm development\n",
    "* Computing a conservation score\n",
    "\n",
    "[Large-Scale Sequence Analysis of Hemagglutinin of Influenza A Virus Identifies Conserved Regions Suitable for Targeting an Anti-Viral Response](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0009268)\n",
    "\n",
    "They are aligning proteins, not RNA.\n",
    "\n",
    "1. full length HA sequences download\n",
    "2. seperate files for each H1, H2 etc\n",
    "3. align each subtype using MUSCLE 3.6 - filtering out frameshifts and partials\n",
    "4. each profile for each multiple alignment then aligned to other profiles using profile-profile aligner COMPASS\n",
    "5. Compute a conservation score for each column of the final alignment\n",
    "\n",
    "[Identification of potential conserved RNA secondary structure throughout influenza A coding regions](https://rnajournal.cshlp.org/content/17/6/991.full_)\n",
    "\n",
    "1. Interesting that this paper aligns first based on protein, then RNA\n",
    "2. six full influenza sets were downloaded from human avian swine\n",
    "3. divide each by segment\n",
    "4. coding regions translated with BioEdit\n",
    "5. protein sequences aligned with ClustalW\n",
    "6. aligned sequences turned back into nt, submitted to RNAz 2.0 for conserved structures\n",
    "\n",
    "Now that I've downloaded all the records that I need and gotten an idea of what they hold, it's time to look at sequences in an attempt to answer question 3: which is how similiar were the recommended vaccine strains to the reported sequences during the 17-18 season. \n",
    "\n",
    "How do I compare the differences between the sequences? Will have to look into BioPython to see what in available in that library to compare genetics. \n",
    "\n",
    "First I'm going to attempt to download the vaccine candidate sequences from NCBI.\n",
    "\n",
    "First one is A/Michigan/45/2015 (H1N1) pdm09-like virus\n",
    "\n",
    "Retrieve that from IRD\n",
    "\n",
    "Some ideas:\n",
    "a script that automatically determines the % identity (essentially a BLAST?) to its matching component in the vaccine.\n",
    "Researcher uploads sequence to database, then an output of all the similiarities in one dataframe, maybe represented in a table\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import data analysis libraries \n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# Import visualization libraries\n",
    "import seaborn as sns\n",
    "sns.set_style('whitegrid')\n",
    "sns.set_context('paper', font_scale=1.)\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Strain Name', 'Complete Genome', 'Subtype', 'Collection Date', 'Host',\n",
       "       'Country', 'State/Province', 'Geographic Grouping', 'Flu Season',\n",
       "       'Submission Date', 'Passage History', 'Specimen Source Health Status',\n",
       "       '1 PB2', '2 PB1', '3 PA', '4 HA', '5 NP', '6 NA', '7 MP', '8 NS', 'Age',\n",
       "       'Gender', 'M2 31N', 'M2 26F', 'M2 27A', 'M2 30T', 'M2 34E',\n",
       "       'NA 275Y N1', 'NA 292K N2', 'NA 119V N2', 'NA 294S N2', 'PB1-F2 66S',\n",
       "       'PB2 E627K', 'PB2 D701N', 'PB2 A199S', 'PB2 A661T', 'PB2 V667I',\n",
       "       'PB2 K702R', 'PA S409N', 'NP L136M', 'M2 A16G', 'M2 C55F', 'NS1 T92E',\n",
       "       'RERRRKKR', 'Sensitive Drug', 'Resistant Drug', 'Submission Date.1',\n",
       "       'NCBI Taxon ID', 'pH1N1-like', 'US Swine H1 Clade',\n",
       "       'Global Swine H1 Clade test', 'H5 Clade', 'Unnamed: 52'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Load up the vaccine candidate sequences into a dataframe, first with H1N1\n",
    "\n",
    "AM42_H1N1_db = pd.read_csv('../A-michigan-45-2015.tsv', sep='\\t')\n",
    "AM42_H1N1_db.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Strain Name</th>\n",
       "      <th>Complete Genome</th>\n",
       "      <th>Subtype</th>\n",
       "      <th>Collection Date</th>\n",
       "      <th>Host</th>\n",
       "      <th>Country</th>\n",
       "      <th>State/Province</th>\n",
       "      <th>Geographic Grouping</th>\n",
       "      <th>Flu Season</th>\n",
       "      <th>Submission Date</th>\n",
       "      <th>...</th>\n",
       "      <th>RERRRKKR</th>\n",
       "      <th>Sensitive Drug</th>\n",
       "      <th>Resistant Drug</th>\n",
       "      <th>Submission Date.1</th>\n",
       "      <th>NCBI Taxon ID</th>\n",
       "      <th>pH1N1-like</th>\n",
       "      <th>US Swine H1 Clade</th>\n",
       "      <th>Global Swine H1 Clade test</th>\n",
       "      <th>H5 Clade</th>\n",
       "      <th>Unnamed: 52</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A/Michigan/45/2015</td>\n",
       "      <td>Yes</td>\n",
       "      <td>H1N1</td>\n",
       "      <td>09/07/2015</td>\n",
       "      <td>Human</td>\n",
       "      <td>USA</td>\n",
       "      <td>Michigan</td>\n",
       "      <td>North America</td>\n",
       "      <td>-N/A-</td>\n",
       "      <td>2017-08-24</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>-N/A-</td>\n",
       "      <td>-N/A-</td>\n",
       "      <td>08/24/2017</td>\n",
       "      <td>1777792</td>\n",
       "      <td>Mixed Positive and Negative Segments</td>\n",
       "      <td>npdm</td>\n",
       "      <td>1A.3.3.2</td>\n",
       "      <td>-N/A-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Strain Name Complete Genome Subtype Collection Date   Host Country  \\\n",
       "0  A/Michigan/45/2015             Yes    H1N1      09/07/2015  Human     USA   \n",
       "\n",
       "  State/Province Geographic Grouping Flu Season Submission Date     ...      \\\n",
       "0       Michigan       North America      -N/A-      2017-08-24     ...       \n",
       "\n",
       "  RERRRKKR Sensitive Drug Resistant Drug Submission Date.1 NCBI Taxon ID  \\\n",
       "0       No          -N/A-          -N/A-        08/24/2017       1777792   \n",
       "\n",
       "                             pH1N1-like US Swine H1 Clade  \\\n",
       "0  Mixed Positive and Negative Segments              npdm   \n",
       "\n",
       "  Global Swine H1 Clade test H5 Clade Unnamed: 52  \n",
       "0                   1A.3.3.2    -N/A-         NaN  \n",
       "\n",
       "[1 rows x 53 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AM42_H1N1_db.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "OK, lets look at just the HA segment - column named '4 HA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    KY117023,KY090610,KU933493,KU509703\n",
       "Name: 4 HA, dtype: object"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AM42_H1N1_db['4 HA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['KY117023', 'KY090610', 'KU933493', 'KU509703']\n"
     ]
    }
   ],
   "source": [
    "#going to put this into a list so can interate over and call into NCBI\n",
    "#listing the contents of a column will yield a series, not a list so need to call .tolist() to convert\n",
    "\n",
    "acc_list = AM42_H1N1_db['4 HA'].tolist()\n",
    "acc_list = acc_list[0].split(',')\n",
    "print(acc_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So this give me the accession numbers in a list. Going iterate over this list to call into NCBI and download these sequences and put this into a database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>FASTA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KY117023</td>\n",
       "      <td>&gt;KY117023.1 Influenza A virus (A/Michigan/45/2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KY090610</td>\n",
       "      <td>&gt;KY090610.1 Influenza A virus (A/Michigan/45/2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KU933493</td>\n",
       "      <td>&gt;KU933493.1 Influenza A virus (A/Michigan/45/2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KU509703</td>\n",
       "      <td>&gt;KU509703.1 Influenza A virus (A/Michigan/45/2...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  accession                                              FASTA\n",
       "0  KY117023  >KY117023.1 Influenza A virus (A/Michigan/45/2...\n",
       "1  KY090610  >KY090610.1 Influenza A virus (A/Michigan/45/2...\n",
       "2  KU933493  >KU933493.1 Influenza A virus (A/Michigan/45/2...\n",
       "3  KU509703  >KU509703.1 Influenza A virus (A/Michigan/45/2..."
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Using the Bio Entrez module, going to retreive the sequence data the four accession nos\n",
    "# Then add this to a new dataframe for storage.\n",
    "\n",
    "from Bio import Entrez\n",
    "\n",
    "Entrez.email = \"adriana@dranalytics.co\"\n",
    "df = pd.DataFrame(columns=['accession', 'FASTA'])\n",
    "\n",
    "\n",
    "for i in range(len(acc_list)):\n",
    "    FASTA = Entrez.efetch(db=\"nucleotide\", id=acc_list[i], rettype=\"fasta\", retmode=\"text\").read()\n",
    "    \n",
    "    df.loc[i] = {'accession': acc_list[i], 'FASTA': FASTA}\n",
    "\n",
    "df.head(4)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now going to run a pairwise sequence alignment - "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pairwise sequence aligner with parameters\n",
      "  match score: 1.000000\n",
      "  mismatch score: 0.000000\n",
      "  target open gap score: 0.000000\n",
      "  target extend gap score: 0.000000\n",
      "  target left open gap score: 0.000000\n",
      "  target left extend gap score: 0.000000\n",
      "  target right open gap score: 0.000000\n",
      "  target right extend gap score: 0.000000\n",
      "  query open gap score: 0.000000\n",
      "  query extend gap score: 0.000000\n",
      "  query left open gap score: 0.000000\n",
      "  query left extend gap score: 0.000000\n",
      "  query right open gap score: 0.000000\n",
      "  query right extend gap score: 0.000000\n",
      "  mode: global\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio import Align\n",
    "aligner = Align.PairwiseAligner()\n",
    "\n",
    "seq1 = \"GAACT\"\n",
    "seq2 = \"GAT\"\n",
    "\n",
    "alignments = aligner.align(seq1, seq2)\n",
    "print(aligner)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ClustalOmegaCommandline',\n",
       " 'ClustalwCommandline',\n",
       " 'DialignCommandline',\n",
       " 'MSAProbsCommandline',\n",
       " 'MafftCommandline',\n",
       " 'MuscleCommandline',\n",
       " 'PrankCommandline',\n",
       " 'ProbconsCommandline',\n",
       " 'TCoffeeCommandline',\n",
       " '_ClustalOmega',\n",
       " '_Clustalw',\n",
       " '_Dialign',\n",
       " '_MSAProbs',\n",
       " '_Mafft',\n",
       " '_Muscle',\n",
       " '_Prank',\n",
       " '_Probcons',\n",
       " '_TCoffee',\n",
       " '__all__',\n",
       " '__builtins__',\n",
       " '__cached__',\n",
       " '__doc__',\n",
       " '__file__',\n",
       " '__loader__',\n",
       " '__name__',\n",
       " '__package__',\n",
       " '__path__',\n",
       " '__spec__']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import Bio.Align.Applications\n",
    "dir(Bio.Align.Applications)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class ClustalOmegaCommandline in module Bio.Align.Applications._ClustalOmega:\n",
      "\n",
      "class ClustalOmegaCommandline(Bio.Application.AbstractCommandline)\n",
      " |  ClustalOmegaCommandline(cmd='clustalo', **kwargs)\n",
      " |  \n",
      " |  Command line wrapper for clustal omega.\n",
      " |  \n",
      " |  http://www.clustal.org/omega\n",
      " |  \n",
      " |  Notes\n",
      " |  -----\n",
      " |  Last checked against version: 1.2.0\n",
      " |  \n",
      " |  References\n",
      " |  ----------\n",
      " |  Sievers F, Wilm A, Dineen DG, Gibson TJ, Karplus K, Li W, Lopez R,\n",
      " |  McWilliam H, Remmert M, Söding J, Thompson JD, Higgins DG (2011).\n",
      " |  Fast, scalable generation of high-quality protein multiple\n",
      " |  sequence alignments using Clustal Omega.\n",
      " |  Molecular Systems Biology 7:539 https://doi.org/10.1038/msb.2011.75\n",
      " |  \n",
      " |  Examples\n",
      " |  --------\n",
      " |  >>> from Bio.Align.Applications import ClustalOmegaCommandline\n",
      " |  >>> in_file = \"unaligned.fasta\"\n",
      " |  >>> out_file = \"aligned.fasta\"\n",
      " |  >>> clustalomega_cline = ClustalOmegaCommandline(infile=in_file, outfile=out_file, verbose=True, auto=True)\n",
      " |  >>> print(clustalomega_cline)\n",
      " |  clustalo -i unaligned.fasta -o aligned.fasta --auto -v\n",
      " |  \n",
      " |  You would typically run the command line with clustalomega_cline() or via\n",
      " |  the Python subprocess module, as described in the Biopython tutorial.\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      ClustalOmegaCommandline\n",
      " |      Bio.Application.AbstractCommandline\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self, cmd='clustalo', **kwargs)\n",
      " |      Initialize the class.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  auto\n",
      " |      Set options automatically (might overwrite some of your options)\n",
      " |      \n",
      " |      This property controls the addition of the --auto switch, treat this property as a boolean.\n",
      " |  \n",
      " |  clusteringout\n",
      " |      Clustering output file\n",
      " |      \n",
      " |      This controls the addition of the --clustering-out parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  clustersize\n",
      " |      soft maximum of sequences in sub-clusters\n",
      " |      \n",
      " |      This controls the addition of the --cluster-size parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  dealign\n",
      " |      Dealign input sequences\n",
      " |      \n",
      " |      This property controls the addition of the --dealign switch, treat this property as a boolean.\n",
      " |  \n",
      " |  distmat_full\n",
      " |      Use full distance matrix for guide-tree calculation (slow; mBed is default)\n",
      " |      \n",
      " |      This property controls the addition of the --full switch, treat this property as a boolean.\n",
      " |  \n",
      " |  distmat_full_iter\n",
      " |      Use full distance matrix for guide-tree calculation during iteration (mBed is default)\n",
      " |      \n",
      " |      This property controls the addition of the --full-iter switch, treat this property as a boolean.\n",
      " |  \n",
      " |  distmat_in\n",
      " |      Pairwise distance matrix input file (skips distance computation).\n",
      " |      \n",
      " |      This controls the addition of the --distmat-in parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  distmat_out\n",
      " |      Pairwise distance matrix output file.\n",
      " |      \n",
      " |      This controls the addition of the --distmat-out parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  force\n",
      " |      Force file overwriting.\n",
      " |      \n",
      " |      This property controls the addition of the --force switch, treat this property as a boolean.\n",
      " |  \n",
      " |  guidetree_in\n",
      " |      Guide tree input file (skips distance computation and guide-tree clustering step).\n",
      " |      \n",
      " |      This controls the addition of the --guidetree-in parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  guidetree_out\n",
      " |      Guide tree output file.\n",
      " |      \n",
      " |      This controls the addition of the --guidetree-out parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  help\n",
      " |      Print help and exit.\n",
      " |      \n",
      " |      This property controls the addition of the -h switch, treat this property as a boolean.\n",
      " |  \n",
      " |  hmm_input\n",
      " |      HMM input files\n",
      " |      \n",
      " |      This controls the addition of the --hmm-in parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  infile\n",
      " |      Multiple sequence input file\n",
      " |      \n",
      " |      This controls the addition of the -i parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  infmt\n",
      " |      Forced sequence input file format (default: auto)\n",
      " |      \n",
      " |                          Allowed values: a2m, fa[sta], clu[stal], msf, phy[lip], selex, st[ockholm], vie[nna]\n",
      " |                          \n",
      " |      \n",
      " |      This controls the addition of the --infmt parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  isprofile\n",
      " |      disable check if profile, force profile (default no)\n",
      " |      \n",
      " |      This property controls the addition of the --is-profile switch, treat this property as a boolean.\n",
      " |  \n",
      " |  iterations\n",
      " |      Number of (combined guide-tree/HMM) iterations\n",
      " |      \n",
      " |      This controls the addition of the --iterations parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  log\n",
      " |      Log all non-essential output to this file.\n",
      " |      \n",
      " |      This controls the addition of the -l parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  long_version\n",
      " |      Print long version information and exit\n",
      " |      \n",
      " |      This property controls the addition of the --long-version switch, treat this property as a boolean.\n",
      " |  \n",
      " |  max_guidetree_iterations\n",
      " |      Maximum number of guidetree iterations\n",
      " |      \n",
      " |      This controls the addition of the --max-guidetree-iterations parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  max_hmm_iterations\n",
      " |      Maximum number of HMM iterations\n",
      " |      \n",
      " |      This controls the addition of the --max-hmm-iterations parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  maxnumseq\n",
      " |      Maximum allowed number of sequences\n",
      " |      \n",
      " |      This controls the addition of the --maxnumseq parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  maxseqlen\n",
      " |      Maximum allowed sequence length\n",
      " |      \n",
      " |      This controls the addition of the --maxseqlen parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  outfile\n",
      " |      Multiple sequence alignment output file (default: stdout).\n",
      " |      \n",
      " |      This controls the addition of the -o parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  outfmt\n",
      " |      MSA output file format: a2m=fa[sta],clu[stal],msf,phy[lip],selex,st[ockholm],vie[nna] (default: fasta).\n",
      " |      \n",
      " |      This controls the addition of the --outfmt parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  outputorder\n",
      " |      MSA output order like in input/guide-tree\n",
      " |      \n",
      " |      This controls the addition of the --output-order parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  percentid\n",
      " |      convert distances into percent identities (default no)\n",
      " |      \n",
      " |      This property controls the addition of the --percent-id switch, treat this property as a boolean.\n",
      " |  \n",
      " |  profile1\n",
      " |      Pre-aligned multiple sequence file (aligned columns will be kept fix).\n",
      " |      \n",
      " |      This controls the addition of the --profile1 parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  profile2\n",
      " |      Pre-aligned multiple sequence file (aligned columns will be kept fix).\n",
      " |      \n",
      " |      This controls the addition of the --profile2 parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  residuenumber\n",
      " |      in Clustal format print residue numbers (default no)\n",
      " |      \n",
      " |      This property controls the addition of the --residuenumber switch, treat this property as a boolean.\n",
      " |  \n",
      " |  seqtype\n",
      " |      {Protein, RNA, DNA} Force a sequence type (default: auto).\n",
      " |      \n",
      " |      This controls the addition of the -t parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  threads\n",
      " |      Number of processors to use\n",
      " |      \n",
      " |      This controls the addition of the --threads parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  usekimura\n",
      " |      use Kimura distance correction for aligned sequences (default no)\n",
      " |      \n",
      " |      This property controls the addition of the --use-kimura switch, treat this property as a boolean.\n",
      " |  \n",
      " |  verbose\n",
      " |      Verbose output\n",
      " |      \n",
      " |      This property controls the addition of the -v switch, treat this property as a boolean.\n",
      " |  \n",
      " |  version\n",
      " |      Print version information and exit\n",
      " |      \n",
      " |      This property controls the addition of the --version switch, treat this property as a boolean.\n",
      " |  \n",
      " |  wrap\n",
      " |      number of residues before line-wrap in output\n",
      " |      \n",
      " |      This controls the addition of the --wrap parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  __call__(self, stdin=None, stdout=True, stderr=True, cwd=None, env=None)\n",
      " |      Execute command, wait for it to finish, return (stdout, stderr).\n",
      " |      \n",
      " |      Runs the command line tool and waits for it to finish. If it returns\n",
      " |      a non-zero error level, an exception is raised. Otherwise two strings\n",
      " |      are returned containing stdout and stderr.\n",
      " |      \n",
      " |      The optional stdin argument should be a string of data which will be\n",
      " |      passed to the tool as standard input.\n",
      " |      \n",
      " |      The optional stdout and stderr argument may be filenames (string),\n",
      " |      but otherwise are treated as a booleans, and control if the output\n",
      " |      should be captured as strings (True, default), or ignored by sending\n",
      " |      it to /dev/null to avoid wasting memory (False). If sent to a file\n",
      " |      or ignored, then empty string(s) are returned.\n",
      " |      \n",
      " |      The optional cwd argument is a string giving the working directory\n",
      " |      to run the command from. See Python's subprocess module documentation\n",
      " |      for more details.\n",
      " |      \n",
      " |      The optional env argument is a dictionary setting the environment\n",
      " |      variables to be used in the new process. By default the current\n",
      " |      process' environment variables are used. See Python's subprocess\n",
      " |      module documentation for more details.\n",
      " |      \n",
      " |      Default example usage::\n",
      " |      \n",
      " |          from Bio.Emboss.Applications import WaterCommandline\n",
      " |          water_cmd = WaterCommandline(gapopen=10, gapextend=0.5,\n",
      " |                                       stdout=True, auto=True,\n",
      " |                                       asequence=\"a.fasta\", bsequence=\"b.fasta\")\n",
      " |          print(\"About to run: %s\" % water_cmd)\n",
      " |          std_output, err_output = water_cmd()\n",
      " |      \n",
      " |      This functionality is similar to subprocess.check_output() added in\n",
      " |      Python 2.7. In general if you require more control over running the\n",
      " |      command, use subprocess directly.\n",
      " |      \n",
      " |      As of Biopython 1.56, when the program called returns a non-zero error\n",
      " |      level, a custom ApplicationError exception is raised. This includes\n",
      " |      any stdout and stderr strings captured as attributes of the exception\n",
      " |      object, since they may be useful for diagnosing what went wrong.\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return a representation of the command line object for debugging.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5)\n",
      " |      >>> cline.asequence = \"asis:ACCCGGGCGCGGT\"\n",
      " |      >>> cline.bsequence = \"asis:ACCCGAGCGCGGT\"\n",
      " |      >>> cline.outfile = \"temp_water.txt\"\n",
      " |      >>> print(cline)\n",
      " |      water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5\n",
      " |      >>> cline\n",
      " |      WaterCommandline(cmd='water', outfile='temp_water.txt', asequence='asis:ACCCGGGCGCGGT', bsequence='asis:ACCCGAGCGCGGT', gapopen=10, gapextend=0.5)\n",
      " |  \n",
      " |  __setattr__(self, name, value)\n",
      " |      Set attribute name to value (PRIVATE).\n",
      " |      \n",
      " |      This code implements a workaround for a user interface issue.\n",
      " |      Without this __setattr__ attribute-based assignment of parameters\n",
      " |      will silently accept invalid parameters, leading to known instances\n",
      " |      of the user assuming that parameters for the application are set,\n",
      " |      when they are not.\n",
      " |      \n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5, stdout=True)\n",
      " |      >>> cline.asequence = \"a.fasta\"\n",
      " |      >>> cline.bsequence = \"b.fasta\"\n",
      " |      >>> cline.csequence = \"c.fasta\"\n",
      " |      Traceback (most recent call last):\n",
      " |      ...\n",
      " |      ValueError: Option name csequence was not found.\n",
      " |      >>> print(cline)\n",
      " |      water -stdout -asequence=a.fasta -bsequence=b.fasta -gapopen=10 -gapextend=0.5\n",
      " |      \n",
      " |      This workaround uses a whitelist of object attributes, and sets the\n",
      " |      object attribute list as normal, for these.  Other attributes are\n",
      " |      assumed to be parameters, and passed to the self.set_parameter method\n",
      " |      for validation and assignment.\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Make the commandline string with the currently set options.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5)\n",
      " |      >>> cline.asequence = \"asis:ACCCGGGCGCGGT\"\n",
      " |      >>> cline.bsequence = \"asis:ACCCGAGCGCGGT\"\n",
      " |      >>> cline.outfile = \"temp_water.txt\"\n",
      " |      >>> print(cline)\n",
      " |      water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5\n",
      " |      >>> str(cline)\n",
      " |      'water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5'\n",
      " |  \n",
      " |  set_parameter(self, name, value=None)\n",
      " |      Set a commandline option for a program (OBSOLETE).\n",
      " |      \n",
      " |      Every parameter is available via a property and as a named\n",
      " |      keyword when creating the instance. Using either of these is\n",
      " |      preferred to this legacy set_parameter method which is now\n",
      " |      OBSOLETE, and likely to be DEPRECATED and later REMOVED in\n",
      " |      future releases.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  parameters = None\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio.Align.Applications import ClustalOmegaCommandline\n",
    "help(ClustalOmegaCommandline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'>KY117023.1 Influenza A virus (A/Michigan/45/2015(H1N1)) segment 4 hemagglutinin (HA) gene, complete cds\\nGGAAAAACAAAAGCAACAAAAATGAAGGCAATACTAGTAGTTCTGCTATATACATTTACAACCGCAAATG\\nCAGACACATTATGTATAGGTTATCATGCGAACAATTCAACAGACACTGTAGACACAGTACTAGAAAAGAA\\nTGTAACAGTAACACACTCTGTTAACCTTCTGGAAGACAAGCATAACGGAAAACTATGCAAACTAAGAGGG\\nGTAGCCCCATTGCATTTGGGTAAATGTAACATTGCTGGCTGGATCCTGGGAAATCCAGAGTGTGAATCAC\\nTCTCCACAGCAAGTTCATGGTCCTACATTGTGGAAACATCTAATTCAGACAATGGAACGTGTTACCCAGG\\nAGATTTCATCAATTATGAGGAGCTAAGAGAGCAATTGAGCTCAGTGTCATCATTTGAAAGGTTTGAGATA\\nTTCCCCAAGACAAGTTCATGGCCCAATCATGACTCGAACAAAGGTGTAACGGCAGCATGTCCTCACGCTG\\nGAGCAAAAAGCTTCTACAAAAACTTGATATGGCTAGTTAAAAAAGGAAATTCATACCCAAAGCTTAACCA\\nATCCTACATTAATGATAAAGGGAAAGAAGTCCTCGTGCTGTGGGGCATTCACCATCCATCTACTACTGCT\\nGACCAACAAAGTCTCTATCAGAATGCAGATGCATATGTTTTTGTGGGGACATCAAGATACAGCAAGAAGT\\nTCAAGCCGGAAATAGCAACAAGACCCAAAGTGAGGGATCAAGAAGGGAGAATGAACTATTACTGGACACT\\nAGTAGAGCCGGGAGACAAAATAACATTCGAAGCAACTGGAAATCTAGTGGTACCGAGATATGCATTCACA\\nATGGAAAGAAATGCTGGATCTGGTATTATCATTTCAGATACACCAGTCCACGATTGCAATACAACTTGTC\\nAGACACCCGAGGGTGCTATAAACACCAGCCTCCCATTTCAGAATATACATCCGATCACAATTGGAAAATG\\nTCCAAAGTATGTAAAAAGCACAAAATTGAGACTGGCCACAGGATTGAGGAATGTTCCGTCTATTCAATCT\\nAGAGGCCTATTCGGGGCCATTGCCGGCTTCATTGAAGGGGGGTGGACAGGGATGGTAGATGGATGGTACG\\nGTTATCACCATCAAAATGAGCAGGGGTCAGGATATGCAGCCGACCTGAAGAGCACACAAAATGCCATTGA\\nCAAGATTACTAACAAAGTAAATTCTGTTATTGAAAAGATGAATACACAGTTCACAGCAGTGGGTAAAGAG\\nTTCAACCACCTGGAAAAAAGAATAGAGAATCTAAATAAAAAAGTTGATGATGGTTTCCTGGACATTTGGA\\nCTTACAATGCCGAACTGTTGGTTCTATTGGAAAATGAAAGAACTTTGGACTATCACGATTCAAATGTGAA\\nGAACTTGTATGAAAAAGTAAGAAACCAGTTAAAAAACAATGCCAAGGAAATTGGAAACGGCTGCTTTGAA\\nTTTTACCACAAATGCGATAACACGTGCATGGAAAGTGTCAAAAATGGGACTTATGACTACCCAAAATACT\\nCAGAGGAAGCAAAATTAAACAGAGAAAAAATAGATGGGGTAAAGCTGGAATCAACAAGGATTTACCAGAT\\nTTTGGCGATCTATTCAACTGTCGCCAGTTCATTGGTACTGGTAGTCTCCCTGGGGGCAATCAGCTTCTGG\\nATGTGCTCTAATGGGTCTCTACAGTGTAGAATATGTATTTAACATTAGGATTTCAGAATCATGAGAAAAA\\nCAC\\n\\n'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['FASTA'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    1885\n",
       "1    1832\n",
       "2    1884\n",
       "3    1832\n",
       "Name: FASTA, dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file = open('1.FASTA', 'w+')\n",
    "\n",
    "def copy(var):\n",
    "    return file.write(var)\n",
    "    file.close()    \n",
    "    \n",
    "df['FASTA'].apply(copy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('Using 1 threads\\nRead 4 sequences (type: DNA) from 1.fa\\nnot more sequences (4) than cluster-size (100), turn off mBed\\nSetting options automatically based on input sequence characteristics (might overwrite some of your options).\\nAuto settings: Enabling mBed.\\nAuto settings: Setting iteration to 1.\\nUsing 3 seeds (chosen with constant stride from length sorted seqs) for mBed (from a total of 4 sequences)\\nCalculating pairwise ktuple-distances...\\nKtuple-distance calculation progress: 0 % (0 out of 9)\\nKtuple-distance calculation progress: 33 % (3 out of 9)\\nKtuple-distance calculation progress: 55 % (5 out of 9)\\nKtuple-distance calculation progress done. CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00\\nmBed created 1 cluster/s (with a minimum of 1 and a soft maximum of 100 sequences each)\\nDistance calculation within sub-clusters: 0 % (0 out of 1)\\nDistance calculation within sub-clusters done. CPU time: 0.02u 0.01s 00:00:00.03 Elapsed: 00:00:00\\nGuide-tree computation (mBed) done.\\nProgressive alignment progress: 33 % (1 out of 3)\\nProgressive alignment progress: 66 % (2 out of 3)\\nProgressive alignment progress: 100 % (3 out of 3)\\nProgressive alignment progress done. CPU time: 1.00u 0.23s 00:00:01.23 Elapsed: 00:00:01\\nIteration step 1 out of 1\\nComputing new guide tree (iteration step 0)\\nCalculating pairwise aligned identity distances...\\nPairwise distance calculation progress: 0 % (0 out of 10)\\nPairwise distance calculation progress: 30 % (3 out of 10)\\nPairwise distance calculation progress: 50 % (5 out of 10)\\nPairwise distance calculation progress: 60 % (6 out of 10)\\nPairwise identity calculation progress done. CPU time: 0.00u 0.00s 00:00:00.00 Elapsed: 00:00:00\\nGuide-tree computation done.\\nComputing HMM from alignment\\nProgressive alignment progress: 33 % (1 out of 3)\\nProgressive alignment progress: 66 % (2 out of 3)\\nProgressive alignment progress: 100 % (3 out of 3)\\nProgressive alignment progress done. CPU time: 3.00u 0.68s 00:00:03.68 Elapsed: 00:00:04\\nAlignment written to aligned.aln\\n',\n",
       " '')"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Align.Applications import ClustalOmegaCommandline\n",
    "in_file = \"1.fa\"\n",
    "out_file = \"aligned.aln\"\n",
    "clustalomega_cline = ClustalOmegaCommandline(infile=in_file, outfile=out_file, verbose=True, auto=True, force=True)\n",
    "clustalomega_cline()\n",
    " \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.FASTA                                 aligned.aln\r\n",
      "1.fa                                    aligned2.FASTA\r\n",
      "Influ_Seq_1205.ipynb                    aligned2.fa\r\n",
      "Influenza_US_Maps.ipynb                 \u001b[1m\u001b[32mfluA_strains.tsv\u001b[m\u001b[m*\r\n",
      "Influenza_inital_data_assessment.ipynb  \u001b[1m\u001b[32mfluB_strains.tsv\u001b[m\u001b[m*\r\n",
      "README.md                               inflA.fasta\r\n",
      "aligned.FASTA                           new_file.fa\r\n"
     ]
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CLUSTAL X (1.81) multiple sequence alignment\n",
      "\n",
      "\n",
      "KY117023.1                          GGAAAAACAAAAGCAACAAAAATGAAGGCAATACTAGTAGTTCTGCTATA\n",
      "KY090610.1                          ---------------------ATGAAGGCAATACTAGTAGTTCTGCTATA\n",
      "KU933493.1                          -GAAAAACAAAAGCAACAAAAATGAAGGCAATACTAGTAGTTCTGCTATA\n",
      "KU509703.1                          ---------------------ATGAAGGCAATACTAGTAGTTCTGCTATA\n",
      "\n",
      "KY117023.1                          TACATTTACAACCGCAAATGCAGACACATTATGTATAGGTTATCATGCGA\n",
      "KY090610.1                          TACATTTACAACCGCAAATGCAGACACATTATGTATAGGTTATCATGCGA\n",
      "KU933493.1                          TACATTTACAACCGCAAATGCAGACACATTATGTATAGGTTATCATGCGA\n",
      "KU509703.1                          TACATTTACAACCGCAAATGCAGACACATTATGTATAGGTTATCATGCGA\n",
      "\n",
      "KY117023.1                          ACAATTCAACAGACACTGTAGACACAGTACTAGAAAAGAATGTAACAGTA\n",
      "KY090610.1                          ACAATTCAACAGACACTGTAGACACAGTACTAGAAAAGAATGTAACAGTA\n",
      "KU933493.1                          ACAATTCAACAGACACTGTAGACACAGTACTAGAAAAGAATGTAACAGTA\n",
      "KU509703.1                          ACAATTCAACAGACACTGTAGACACAGTACTAGAAAAGAATGTAACAGTA\n",
      "\n",
      "KY117023.1                          ACACACTCTGTTAACCTTCTGGAAGACAAGCATAACGGAAAACTATGCAA\n",
      "KY090610.1                          ACACACTCTGTTAACCTTCTGGAAGACAAGCATAACGGAAAACTATGCAA\n",
      "KU933493.1                          ACACACTCTGTTAACCTTCTGGAAGACAAGCATAACGGAAAACTATGCAA\n",
      "KU509703.1                          ACACACTCTGTTAACCTTCTGGAAGACAAGCATAACGGAAAACTATGCAA\n",
      "\n",
      "KY117023.1                          ACTAAGAGGGGTAGCCCCATTGCATTTGGGTAAATGTAACATTGCTGGCT\n",
      "KY090610.1                          ACTAAGAGGGGTAGCCCCATTGCATTTGGGTAAATGTAACATTGCTGGCT\n",
      "KU933493.1                          ACTAAGAGGGGTAGCCCCATTGCATTTGGGTAAATGTAACATTGCTGGCT\n",
      "KU509703.1                          ACTAAGAGGGGTAGCCCCATTGCATTTGGGTAAATGTAACATTGCTGGCT\n",
      "\n",
      "KY117023.1                          GGATCCTGGGAAATCCAGAGTGTGAATCACTCTCCACAGCAAGTTCATGG\n",
      "KY090610.1                          GGATCCTGGGAAATCCAGAGTGTGAATCACTCTCCACAGCAAGTTCATGG\n",
      "KU933493.1                          GGATCCTGGGAAATCCAGAGTGTGAATCACTCTCCACAGCAAGTTCATGG\n",
      "KU509703.1                          GGATCCTGGGAAATCCAGAGTGTGAATCACTCTCCACAGCAAGTTCATGG\n",
      "\n",
      "KY117023.1                          TCCTACATTGTGGAAACATCTAATTCAGACAATGGAACGTGTTACCCAGG\n",
      "KY090610.1                          TCCTACATTGTGGAAACATCTAATTCAGACAATGGAACGTGTTACCCAGG\n",
      "KU933493.1                          TCCTACATTGTGGAAACATCTAATTCAGACAATGGAACGTGTTACCCAGG\n",
      "KU509703.1                          TCCTACATTGTGGAAACATCTAATTCAGACAATGGAACGTGTTACCCAGG\n",
      "\n",
      "KY117023.1                          AGATTTCATCAATTATGAGGAGCTAAGAGAGCAATTGAGCTCAGTGTCAT\n",
      "KY090610.1                          AGATTTCATCAATTATGAGGAGCTAAGAGAGCAATTGAGCTCAGTGTCAT\n",
      "KU933493.1                          AGATTTCATCAATTATGAGGAGCTAAGAGAGCAATTGAGCTCAGTGTCAT\n",
      "KU509703.1                          AGATTTCATCAATTATGAGGAGCTAAGAGAGCAATTGAGCTCAGTGTCAT\n",
      "\n",
      "KY117023.1                          CATTTGAAAGGTTTGAGATATTCCCCAAGACAAGTTCATGGCCCAATCAT\n",
      "KY090610.1                          CATTTGAAAGGTTTGAGATATTCCCCAAGACAAGTTCATGGCCCAATCAT\n",
      "KU933493.1                          CATTTGAAAGGTTTGAGATATTCCCCAAGACAAGTTCATGGCCCAATCAT\n",
      "KU509703.1                          CATTTGAAAGGTTTGAGATATTCCCCAAGACAAGTTCATGGCCCAATCAT\n",
      "\n",
      "KY117023.1                          GACTCGAACAAAGGTGTAACGGCAGCATGTCCTCACGCTGGAGCAAAAAG\n",
      "KY090610.1                          GACTCGAACAAAGGTGTAACGGCAGCATGTCCTCACGCTGGAGCAAAAAG\n",
      "KU933493.1                          GACTCGAACAAAGGTGTAACGGCAGCATGTCCTCACGCTGGAGCAAAAAG\n",
      "KU509703.1                          GASTCGAACAAAGGTGTAACGGCAGCATGTCCTCACGCTGGAGCAAAAAG\n",
      "\n",
      "KY117023.1                          CTTCTACAAAAACTTGATATGGCTAGTTAAAAAAGGAAATTCATACCCAA\n",
      "KY090610.1                          CTTCTACAAAAACTTGATATGGCTAGTTAAAAAAGGAAATTCATACCCAA\n",
      "KU933493.1                          CTTCTACAAAAACTTGATATGGCTAGTTAAAAAAGGAAATTCATACCCAA\n",
      "KU509703.1                          CTTCTACAAAAACTTGATATGGCTAGTTAAAAAAGGAAATTCATACCCAA\n",
      "\n",
      "KY117023.1                          AGCTTAACCAATCCTACATTAATGATAAAGGGAAAGAAGTCCTCGTGCTG\n",
      "KY090610.1                          AGCTTAACCAATCCTACATTAATGATAAAGGGAAAGAAGTCCTCGTGCTG\n",
      "KU933493.1                          AGCTTAACCAATCCTACATTAATGATAAAGGGAAAGAAGTCCTCGTGCTG\n",
      "KU509703.1                          AGCTTAACCAATCCTACATTAATGATAAAGGGAAAGAAGTCCTCGTGCTG\n",
      "\n",
      "KY117023.1                          TGGGGCATTCACCATCCATCTACTACTGCTGACCAACAAAGTCTCTATCA\n",
      "KY090610.1                          TGGGGCATTCACCATCCATCTACTACTGCTGACCAACAAAGTCTCTATCA\n",
      "KU933493.1                          TGGGGCATTCACCATCCATCTACTACTGCTGACCAACAAAGTCTCTATCA\n",
      "KU509703.1                          TGGGGCATTCACCATCCATCTACTACTGCTGACCAACAAAGTMTCTATCA\n",
      "\n",
      "KY117023.1                          GAATGCAGATGCATATGTTTTTGTGGGGACATCAAGATACAGCAAGAAGT\n",
      "KY090610.1                          GAATGCAGATGCATATGTTTTTGTGGGGACATCAAGATACAGCAAGAAGT\n",
      "KU933493.1                          GAATGCAGATGCATATGTTTTTGTGGGGACATCAAGATACAGCAAGAAGT\n",
      "KU509703.1                          GAATGCAGATGCATATGTTTTTGTGGGGACATCAAGATACAGCAAGAAGT\n",
      "\n",
      "KY117023.1                          TCAAGCCGGAAATAGCAACAAGACCCAAAGTGAGGGATCAAGAAGGGAGA\n",
      "KY090610.1                          TCAAGCCGGAAATAGCAACAAGACCCAAAGTGAGGGATCAAGAAGGGAGA\n",
      "KU933493.1                          TCAAGCCGGAAATAGCAACAAGACCCAAAGTGAGGGATCGAGAAGGGAGA\n",
      "KU509703.1                          TCAAGCCGGAAATAGCAACAAGACCCAAAGTGAGGGATCAAGAAGGGAGA\n",
      "\n",
      "KY117023.1                          ATGAACTATTACTGGACACTAGTAGAGCCGGGAGACAAAATAACATTCGA\n",
      "KY090610.1                          ATGAACTATTACTGGACACTAGTAGAGCCGGGAGACAAAATAACATTCGA\n",
      "KU933493.1                          ATGAACTATTACTGGACACTAGTAGAGCCGGGAGACAAAATAACATTCGA\n",
      "KU509703.1                          ATGAACTATTACTGGACACTAGTAGAGCCGGGAGACAAAATAACATTCGA\n",
      "\n",
      "KY117023.1                          AGCAACTGGAAATCTAGTGGTACCGAGATATGCATTCACAATGGAAAGAA\n",
      "KY090610.1                          AGCAACTGGAAATCTAGTGGTACCGAGATATGCATTCACAATGGAAAGAA\n",
      "KU933493.1                          AGCAACTGGAAATCTAGTGGTACCGAGATATGCATTCACAATGGAAAGAA\n",
      "KU509703.1                          AGCAACTGGAAATCTAGTGGTACCGAGATATGCATTCACAATGGAAAGAA\n",
      "\n",
      "KY117023.1                          ATGCTGGATCTGGTATTATCATTTCAGATACACCAGTCCACGATTGCAAT\n",
      "KY090610.1                          ATGCTGGATCTGGTATTATCATTTCAGATACACCAGTCCACGATTGCAAT\n",
      "KU933493.1                          ATGCTGGATCTGGTATTATCATTTCAGATACACCAGTCCACGATTGCAAT\n",
      "KU509703.1                          ATGCTGGATCTGGTATTATCATTTCAGATACACCAGTCCACGATTGCAAT\n",
      "\n",
      "KY117023.1                          ACAACTTGTCAGACACCCGAGGGTGCTATAAACACCAGCCTCCCATTTCA\n",
      "KY090610.1                          ACAACTTGTCAGACACCCGAGGGTGCTATAAACACCAGCCTCCCATTTCA\n",
      "KU933493.1                          ACAACTTGTCAGACACCCGAGGGTGCTATAAACACCAGCCTCCCATTTCA\n",
      "KU509703.1                          ACAACTTGTCAGACACCCGAGGGTGCTATAAACACCAGCCTCCCATTTCA\n",
      "\n",
      "KY117023.1                          GAATATACATCCGATCACAATTGGAAAATGTCCAAAGTATGTAAAAAGCA\n",
      "KY090610.1                          GAATATACATCCGATCACAATTGGAAAATGTCCAAAGTATGTAAAAAGCA\n",
      "KU933493.1                          GAATATACATCCGATCACAATTGGAAAATGTCCAAAGTATGTAAAAAGCA\n",
      "KU509703.1                          GAATATACATCCGATCACAATTGGAAAATGTCCAAAGTATGTAAAAAGCA\n",
      "\n",
      "KY117023.1                          CAAAATTGAGACTGGCCACAGGATTGAGGAATGTTCCGTCTATTCAATCT\n",
      "KY090610.1                          CAAAATTGAGACTGGCCACAGGATTGAGGAATGTTCCGTCTATTCAATCT\n",
      "KU933493.1                          CAAAATTGAGACTGGCCACAGGATTGAGGAATGTTCCGTCTATTCAATCT\n",
      "KU509703.1                          CAAAATTGAGACTGGCCACAGGATTGAGGAATGTTCCGTCTATTCAATCT\n",
      "\n",
      "KY117023.1                          AGAGGCCTATTCGGGGCCATTGCCGGCTTCATTGAAGGGGGGTGGACAGG\n",
      "KY090610.1                          AGAGGCCTATTCGGGGCCATTGCCGGCTTCATTGAAGGGGGGTGGACAGG\n",
      "KU933493.1                          AGAGGCCTATTCGGGGCCATTGCCGGCTTCATTGAAGGGGGGTGGACAGG\n",
      "KU509703.1                          AGAGGCCTATTCGGGGCCATTGCCGGCTTCATTGAAGGGGGGTGGACAGG\n",
      "\n",
      "KY117023.1                          GATGGTAGATGGATGGTACGGTTATCACCATCAAAATGAGCAGGGGTCAG\n",
      "KY090610.1                          GATGGTAGATGGATGGTACGGTTATCACCATCAAAATGAGCAGGGGTCAG\n",
      "KU933493.1                          GATGGTAGATGGATGGTACGGTTATCACCATCAAAATGAGCAGGGGTCAG\n",
      "KU509703.1                          GATGGTAGATGGATGGTACGGTTATCACCATCAAAATGAGCAGGGGTCAG\n",
      "\n",
      "KY117023.1                          GATATGCAGCCGACCTGAAGAGCACACAAAATGCCATTGACAAGATTACT\n",
      "KY090610.1                          GATATGCAGCCGACCTGAAGAGCACACAAAATGCCATTGACAAGATTACT\n",
      "KU933493.1                          GATATGCAGCCGACCTGAAGAGCACACAAAATGCCATTGACAAGATTACT\n",
      "KU509703.1                          GATATGCAGCCGACCTGAAGAGCACACAAAATGCCATTGACAAGATTACT\n",
      "\n",
      "KY117023.1                          AACAAAGTAAATTCTGTTATTGAAAAGATGAATACACAGTTCACAGCAGT\n",
      "KY090610.1                          AACAAAGTAAATTCTGTTATTGAAAAGATGAATACACAGTTCACAGCAGT\n",
      "KU933493.1                          AACAAAGTAAATTCTGTTATTGAAAAGATGAATACACAGTTCACAGCAGT\n",
      "KU509703.1                          AACAAAGTAAATTCTGTTATTGAAAAGATGAATACACAGTTCACAGCAGT\n",
      "\n",
      "KY117023.1                          GGGTAAAGAGTTCAACCACCTGGAAAAAAGAATAGAGAATCTAAATAAAA\n",
      "KY090610.1                          GGGTAAAGAGTTCAACCACCTGGAAAAAAGAATAGAGAATCTAAATAAAA\n",
      "KU933493.1                          GGGTAAAGAGTTCAACCACCTGGAAAAAAGAATAGAGAATCTAAATAAAA\n",
      "KU509703.1                          GGGTAAAGAGTTCAACCACCTGGAAAAAAGAATAGAGAATCTAAATAAAA\n",
      "\n",
      "KY117023.1                          AAGTTGATGATGGTTTCCTGGACATTTGGACTTACAATGCCGAACTGTTG\n",
      "KY090610.1                          AAGTTGATGATGGTTTCCTGGACATTTGGACTTACAATGCCGAACTGTTG\n",
      "KU933493.1                          AAGTTGATGATGGTTTCCTGGACATTTGGACTTACAATGCCGAACTGTTG\n",
      "KU509703.1                          AAGTTGATGATGGTTTCCTGGACATTTGGACTTACAATGCCGAACTGTTG\n",
      "\n",
      "KY117023.1                          GTTCTATTGGAAAATGAAAGAACTTTGGACTATCACGATTCAAATGTGAA\n",
      "KY090610.1                          GTTCTATTGGAAAATGAAAGAACTTTGGACTATCACGATTCAAATGTGAA\n",
      "KU933493.1                          GTTCTATTGGAAAATGAAAGAACTTTGGACTATCACGATTCAAATGTGAA\n",
      "KU509703.1                          GTTCTATTGGAAAATGAAAGAACTTTGGACTATCACGATTCAAATGTGAA\n",
      "\n",
      "KY117023.1                          GAACTTGTATGAAAAAGTAAGAAACCAGTTAAAAAACAATGCCAAGGAAA\n",
      "KY090610.1                          GAACTTGTATGAAAAAGTAAGAAACCAGTTAAAAAACAATGCCAAGGAAA\n",
      "KU933493.1                          GAACTTGTATGAAAAAGTAAGAAACCAGTTAAAAAACAATGCCAAGGAAA\n",
      "KU509703.1                          GAACTTGTATGAAAAAGTAAGAAACCAGTTAAAAAACAATGCCAAGGAAA\n",
      "\n",
      "KY117023.1                          TTGGAAACGGCTGCTTTGAATTTTACCACAAATGCGATAACACGTGCATG\n",
      "KY090610.1                          TTGGAAACGGCTGCTTTGAATTTTACCACAAATGCGATAACACGTGCATG\n",
      "KU933493.1                          TTGGAAACGGCTGCTTTGAATTTTACCACAAATGCGATAACACGTGCATG\n",
      "KU509703.1                          TTGGAAACGGCTGCTTTGAATTTTACCACAAATGCGATAACACGTGCATG\n",
      "\n",
      "KY117023.1                          GAAAGTGTCAAAAATGGGACTTATGACTACCCAAAATACTCAGAGGAAGC\n",
      "KY090610.1                          GAAAGTGTCAAAAATGGGACTTATGACTACCCAAAATACTCAGAGGAAGC\n",
      "KU933493.1                          GAAAGTGTCAAAAATGGGACTTATGACTACCCAAAATACTCAGAGGAAGC\n",
      "KU509703.1                          GAAAGTGTCAAAAATGGGACTTATGACTACCCAAAATACTCAGAGGAAGC\n",
      "\n",
      "KY117023.1                          AAAATTAAACAGAGAAAAAATAGATGGGGTAAAGCTGGAATCAACAAGGA\n",
      "KY090610.1                          AAAATTAAACAGAGAAAAAATAGATGGGGTAAAGCTGGAATCAACAAGGA\n",
      "KU933493.1                          AAAATTAAACAGAGAAAAAATAGATGGGGTAAAGCTGGAATCAACAAGGA\n",
      "KU509703.1                          AAAATTAAACAGAGAAAAAATAGATGGGGTAAAGCTGGAATCAACAAGGA\n",
      "\n",
      "KY117023.1                          TTTACCAGATTTTGGCGATCTATTCAACTGTCGCCAGTTCATTGGTACTG\n",
      "KY090610.1                          TTTACCAGATTTTGGCGATCTATTCAACTGTCGCCAGTTCATTGGTACTG\n",
      "KU933493.1                          TTTACCAGATTTTGGCGATCTATTCAACTGTCGCCAGTTCATTGGTACTG\n",
      "KU509703.1                          TTTACCAGATTTTGGCGATCTATTCAACTGTCGCCAGTTCATTGGTACTG\n",
      "\n",
      "KY117023.1                          GTAGTCTCCCTGGGGGCAATCAGCTTCTGGATGTGCTCTAATGGGTCTCT\n",
      "KY090610.1                          GTAGTCTCCCTGGGGGCAATCAGCTTCTGGATGTGCTCTAATGGGTCTCT\n",
      "KU933493.1                          GTAGTCTCCCTGGGGGCAATCAGCTTCTGGATGTGCTCTAATGGGTCTCT\n",
      "KU509703.1                          GTAGTCTCCCTGGGGGCAATCAGCTTCTGGATGTGCTCTAATGGGTCTCT\n",
      "\n",
      "KY117023.1                          ACAGTGTAGAATATGTATTTAACATTAGGATTTCAGAATCATGAGAAAAA\n",
      "KY090610.1                          ACAGTGTAGAATATGTATTTAA----------------------------\n",
      "KU933493.1                          ACAGTGTAGAATATGTATTTAACATTAGGATTTCAGAATCATGAGAAAAA\n",
      "KU509703.1                          ACAGTGTAGAATATGTATTTAA----------------------------\n",
      "\n",
      "KY117023.1                          CAC\n",
      "KY090610.1                          ---\n",
      "KU933493.1                          CAC\n",
      "KU509703.1                          ---\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "\n",
    "align = AlignIO.read(open('aligned.aln'), 'fasta')\n",
    "print(align.format('clustal'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
